contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,204 @@
+
+## Design 4: Hybrid RPC-CRD Approach
+Keep the same CRD object described in Design 3, but only master has write access to them. Workers must persist its delete state locally and report its status to master through gRPC. Master makes requests to workers by writing to CRD, and fetches worker status by polling (similar to Design 1).
+
+**TODO**: Use network policy to limit pod communication to only within the provisioner system.
+
+### Considerations
+* It's not safe against man-in-the-middle attacks as is. Requires encryption.
+
+## Design 5: Worker as Jobs, Watcher in DaemonSet
+
+### Overview
+One of the major issues with previous approaches is worker identity - when a worker tries to communicate with master, it has to prove it's the worker it claims to be. To fix these issues, each worker must maintain different credentials.
+
+Alternatively, master can actively create worker Jobs to perform discovery and delete rather than having long-running worker pods in a DaemonSet sending data to the master. To trigger discovery, we also have local volume directory watcher pods running in a DaemonSet.
+
+### Worker
+Workers are spawned as Jobs by master on demand. In discovery mode, a worker writes the list of local volumes in JSON format to a file inside an emptyDir. A sidecar container then reads the file and writes to its STDOUT. In deletion mode, the worker executes async delete and uses the status of the Job as an indication of deletion state.
+
+### Watcher
+Watchers are deployed as a DaemonSet. Each watcher keeps an in-memory cache of local volumes and periodically checks the volume directory for changes. If there is a change, it notifies master through gRPC.
+
+If there is a network failure, the gRPC call will fail and the watcher retries with exponential backoff. If master fails after receiving the notification, once it comes back up it will spawn a discovery worker on every node (as explained in the Master section).
+
+### Master
+Master exposes the same RPC calls as Design 1 to enable discovery notifications.
+The master provisioner exposes the following RPC:
+| RPC                           | Description |
+| --- | --- |
+| `LocalVolumeReady()`             | Notifies master that there are new local volumes. Also called on watcher initialization. |
+
+Master can be made completely event driven using Job informers and watcher notification. Master deploys a discovery worker when it gets a notification from a watcher. When the Job is complete, master parses the output of the sidecar container as the list of discovered volumes. If the output is empty (may occur if container logs have been rotated) or invalid, it must re-deploy the worker. Master deploys a deletion worker when a PVC is released and uses the Job's status as deletion state. When the Job is complete, it removes the corresponding PV's finalizer and the PV resource itself.
```

Comment:
I would suggest having a nomenclature for the worker Jobs. For cleanup it could be
<operation name>_nodename_pvname
For example: cleanup_minion1.acme.com_lv-1234
For discover: discover_minion1.acme.com

The advantage of this naming is that kubernetes will automatically provide some kind of single threaded behavior here. You will not be able to create a second job of the same name, until the first job has been deleted.  I see some value in this single threaded behavior for cleanup workers (having more than one active for the same PV is undesirable). For discovery, this is less important. The con is that one has to delete terminated jobs, not sure if there is anything undesirable about doing that.

Reply:
If a job has an error, we might want to keep it around so the container logs aren't evicted

Reply:
Ah OK, in that case I guess we need to keep the terminated jobs.

Reply:
The nomenclature can be augmented to have an integer prefix that increments when error jobs exist. For example, if two jobs in error state exist named `cleanup_0_minion1.acme.com_lv-1234` and `cleanup_1_minion1.acme.com_lv-1234`, the next created job would be `cleanup_2_minion1.acme.com_lv-1234`. In the common case (no errors) we still get the single-threaded benefit.